### ISSUE TITLE
JSON support

### ISSUE THREAD

**Issue by jeremad** [GoogleCloudPlatform/ops-agent#209]
From my understanding of the [documentation](https://cloud.google.com/monitoring/agent/ops-agent/configuration#logging-processors), there should be a support for JSON logs.

Logs from docker containers look like this:
```json
{"log":"some log","stream":"stdout","time":"2021-09-20T14:06:10.422145949Z"}
```

My configuration file is the following:
```yaml
---
logging:
  receivers:
    docker:
      type: "files"
      include_paths:
        - "/var/lib/docker/containers/*/*-json.log"
  processors:
    docker:
      type: "parse_json"
      field: "log"
      time_key: "time"
      time_format: "%Y-%m-%dT%H:%M:%S.%L%Z"
  service:
    pipelines:
      docker:
        receivers: [docker]
        processors: [docker]
metrics:
  service:
    pipelines:
      default_pipeline:
        receivers: []
```

I assumed this would produce, either a structured log entry in `jsonPayload`, or only extract the field I put in the configuration (`log` and `time`), but all I see is my raw json:

Expected:
```json
{
  "jsonPayload": {
    "message": "some log",
   },
  "timestamp": "2021-09-20T14:06:11.150181996Z"
}
```
OR
```json
{
  "jsonPayload": {
    "log": "some log",
    "time": "2021-09-20T14:06:11.150181996Z",
    "stream":"stdout"
  }
}
```

Observed:
```json
{
  "jsonPayload": {
    "message": "{\"log\":\"some log\",\"stream\":\"stdout\",\"time\":\"2021-09-20T14:06:10.422145949Z\"}"
  }
}
```

Have I missed something?
Is there any way to achieve what I want?

**Comment by igorpeshansky** [comment]
The "files" receiver produces a record with a "message" field that contains the entire log line (in this case, a serialized JSON message produced by the Docker log driver). The "field" parameter of the "parse_json" processor controls which field of the log record should be parsed as JSON, and defaults to "message" (to match the behavior of the various built-in receivers). By setting the "field" parameter to "log" in your original config, you were telling "parse_json" to look for the "log" field in the record — but the log record only contained the "message" field, so the parser gave up and left the original record alone. Your colleague's config did not specify "field", which kept the default value ("message"), and thus parsed the record correctly. If this wasn't apparent from our documentation, we would love suggestions on improving it to make that clearer.

Note that your colleague's config did not extract the timestamp from the log record (it was retained as a field in the `jsonPayload` instead). To actually extract it, you can keep the "time_key" and "time_format" parameters from your original config, which should pull out the value as the timestamp of the resulting log entry.

Hope this explains what you observed. Please reopen if you still experience the issue after the above corrections.

**Comment by jeremad** [comment]
Thank you, It makes much more sense now.

Here is how I would rephrase the documentation:

Depending on the value of the type element, there are other configuration options, as follows:

* Both parse_json and parse_regex processors:
  * time_key: Optional. If the log entry provides a field with a timestamp, this option specifies the name of that field. The extracted value is used to set the timestamp field of the resulting LogEntry and is removed from the payload..

    If the time_key option is specified, you must also specify the following:

    * time_format: Required if time_key is used. This option specifies the format of the time_key field so it can be recognized and analyzed properly. For details of the format, see the strptime(3) guide.
 
* parse_regex processors:
  * field: Optional. The name of the field in the record to parse. If the field option isn't specified, then the processor parses the message field.
  * regex: Required. The regular expression for parsing the field. The expression must include key names for the matched subexpressions; for example, "^(?<time>[^ ]*) (?<severity>[^ ]*) (?<msg>.*)$".

* parse_json processors:
  * field: Optional. The name of the field in the record to parse as JSON. If the field option isn't specified, then the processor parses the message field. This is the default for built-in receivers, including `file`.

For a set of regular expressions for extracting information from log files, see Common log files.

**Comment by alanhala** [comment]
@jeremad I think if you want to achieve what you describe in your issue you have to apply this configuration:
```yaml
parse-json:
  type: parse_json
  time_key: time
  time_format: '%Y-%m-%dT%H:%M:%S.%LZ'
parse-log:
  type: parse_json
  field: log
```
The first processor would just parse the log from docker, returning a json with log, stream and time. The second processor would take whatever is in the log field and would parse it as a json again, and that would be the log you are expecting.
